compositor_pipeline/src/pipeline/decoder.rs

```diff
@@ -2,33 +2,142 @@ use std::sync::Arc;
 
 use crate::{error::DecoderInitError, queue::Queue};
 
-use self::ffmpeg_h264::H264FfmpegDecoder;
+use self::{ffmpeg_h264::H264FfmpegDecoder, opus_decoder::OpusDecoder};
 
-use super::structs::EncodedChunk;
+use super::{
+    input::rtp::ChunksReceiver,
+    structs::{AudioChannels, AudioCodec, EncodedChunk, VideoCodec},
+};
 use compositor_render::InputId;
+use crossbeam_channel::Receiver;
 
 pub mod ffmpeg_h264;
+mod opus_decoder;
 
 pub enum Decoder {
-    H264(H264FfmpegDecoder),
+    Video(VideoDecoder),
+    Audio(AudioDecoder),
+    VideoWithAudio {
+        video: VideoDecoder,
+        audio: AudioDecoder,
+    },
+}
+
+#[derive(Debug, Clone)]
+pub enum DecoderOptions {
```

Comment:
This seems like an unnecessary complication, you could just represent it as 
```
struct {
  video: Option<VideoDecoder>,
  audio: Option<AudioDecoder>,
}
```

I suspect your goal was to disallow the representation of an invalid state. Normally we want to do that to avoid maintaining additional logic to handle edge cases that those states present, but here your choice has the opposite effect. Stream without video and audio is perfectly valid and wouldn't require additional maintenance, we want to disallow it on the validation layer, but there is no reason to block it in internal abstraction. By selecting an enum you are forcing yourself to handle 3 cases everywhere instead of 2